xc/device.ml

```diff
@@ -1983,30 +1983,52 @@ module Dm_Common = struct
     ; fd_map = []
     }
 
+  let nvidia_compat_lookup config_file =
+    let rec find = function
+      | [] -> raise Not_found
+      | conf :: others ->
+        match String.split_on_char ' ' conf with
+        | [cf; type_id] when cf = config_file -> type_id
+        | _ -> find others
+    in
+    try
+      Unixext.string_of_file !Xenopsd.nvidia_compat_lookup_file
+      |> String.split_on_char '\n'
+      |> find
+    with e ->
+      error "nvidia_compat_lookup failed with %s" (Printexc.to_string e);
+      raise (Xenopsd_error (Internal_error (Printf.sprintf "NVidia vGPU compat metadata lookup failed (%s)" __LOC__)))
+
   let vgpu_args_of_nvidia domid vcpus vgpus restore =
     let open Xenops_interface.Vgpu in
     let virtual_pci_address_compare vgpu1 vgpu2 =
       match vgpu1, vgpu2 with
-      |{implementation = Nvidia {virtual_pci_address = Some pci1; _}; _},
-       {implementation = Nvidia {virtual_pci_address = Some pci2; _}; _} ->
+      |{implementation = Nvidia {virtual_pci_address = pci1; _}; _},
+       {implementation = Nvidia {virtual_pci_address = pci2; _}; _} ->
         Pervasives.compare pci1.dev pci2.dev
       | other1, other2 -> Pervasives.compare other1 other2
       in
-    let get_pci_string = function
-      | None -> raise (Xenopsd_error (Internal_error (Printf.sprintf "No PCI address (%s)" __LOC__)))
-      | Some pci -> Xcp_pci.string_of_address pci in
     let device_args =
       List.map (fun x ->
-          match x.implementation with
-          | Nvidia _conf ->
+          let make args =
             Printf.sprintf "--device=%s"
-              ([(Xenops_interface.Pci.string_of_address x.physical_pci_address);
-                _conf.type_id;
-                (get_pci_string _conf.virtual_pci_address);
-                _conf.uuid;
-                _conf.extra_args]
+              (Xcp_pci.string_of_address x.physical_pci_address :: args
                |> List.filter (fun str -> str <> "")
                |> String.concat ",")
+          in
+          match x.implementation with
+          | Nvidia {virtual_pci_address; type_id = Some type_id; uuid = Some uuid; extra_args} ->
+            make [type_id; Xcp_pci.string_of_address virtual_pci_address; uuid; extra_args]
+          | Nvidia {virtual_pci_address; config_file = Some config_file; extra_args} ->
+            (* Upgrade case: find the type_id that match the give config_file path. *)
+            let type_id = nvidia_compat_lookup config_file in
+            debug "NVidia vGPU config: matched config_file=%s with type_id=%s" config_file type_id;
+            (* The VGPU UUID is not available. Filling in zeros; xapi will deal with it. *)
+            let uuid = "00000000-0000-0000-0000-000000000000" in
```

Comment:
you may also want to change https://github.com/xenserver/gpumon/blob/master/gpumon/gpumon_server.ml#L77 
to something like ` | ""|"00000000-0000-0000-0000-000000000000" -> List.map (fun vgpu -> vgpu)`
when gppumon try to find a vgpu for a VM from a pgpu, `00000000-0000-0000-0000-000000000000` now means the only vGPup for the VM.

Reply:
My understanding is `"00000000-0000-0000-0000-000000000000"` will be passed to demu as part of the `--device` args, then this will be set as the id of the vGPU (Nvidia would set it as the id of the vGPU).
If there are multiple VM starts in this case, then there would be multiple vGPUs use such ID, would it cause some issue?

Reply:
I prefer to avoid any dependency on "special values" as much as possible. In this patch, the all-zero UUID doesn't mean anything (it could have any other string), and it is effectively unused. We don't need in order to differentiate between multiple VGPUs, because there will be just one per VM for the cases where this is used.

To your second point, xapi and gpumon first use the domid to find the VGPUs for a VM. So there is no ambiguity.

Reply:
To avoid any possible issues around the UUID not being unique, however, I will change it from zeros to a random new UUID.

Reply:
Updated the latter.

Reply:
https://github.com/xenserver/gpumon/blob/master/gpumon/gpumon_server.ml#L78 will find a vGPU for the VM from a phisical pci bus with a given uuid, (the uuid is from xapi), in this case, xenopsd set a different uuid with xapi, then gpupmon may cannot find the vgpu device.

Reply:
@liulinC I've handled that in xapi: https://github.com/xapi-project/xen-api/pull/3890/commits/3e67a50a8c33834de11c126f573464c8876c8588